src/lock_utilities.rs

```diff
@@ -0,0 +1,200 @@
+use parking_lot::lock_api::{ArcRwLockReadGuard, ArcRwLockWriteGuard, RawRwLock as RawRwLockTrait};
+use parking_lot::{Mutex, MutexGuard, RawRwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::fmt::Debug;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+
+use crate::memory_region::local::LocalMrInner;
+
+/// Used to transfer rwlock read guard to another type
+pub struct MappedRwLockReadGuard<'a, T> {
```

Comment:
`parking_lot` has the guards, why you need to rewrite them?

Reply:
Because we need `F: FnOnce(T) -> U` here.

`parking_lot` has :
```
 map<F, U>(s: Self, f: F) -> MappedRwLockReadGuard<'a, U>
    where
        F: FnOnce(&T) -> &U,
```
We need:
```
 map<F, U>(s: Self, f: F) -> MappedRwLockReadGuard<'a, U>
    where
        F: FnOnce(T) -> U,
```